backend/proxy/secure_socks_proxy.go

```diff
@@ -236,3 +252,64 @@ func SecureSocksProxyEnabledOnDS(jsonData map[string]interface{}) bool {
 
 	return false
 }
+
+// instrumentedSocksDialer  is a wrapper around the proxy.Dialer and proxy.DialContext
+// that records relevant socks secure socks proxy.
+type instrumentedSocksDialer struct {
+	dialer proxy.Dialer
+}
+
+// NewInstrumenSockstedDialer creates a new instrumented dialer
+func NewInstrumentedSocksDialer(dialer proxy.Dialer) proxy.Dialer {
+	return &instrumentedSocksDialer{
+		dialer: dialer,
+	}
+}
+
+// Dial -
+func (d *instrumentedSocksDialer) Dial(network, addr string) (net.Conn, error) {
+	return d.DialContext(context.Background(), network, addr)
+}
+
+// DialContext -
+func (d *instrumentedSocksDialer) DialContext(ctx context.Context, n, addr string) (net.Conn, error) {
+	start := time.Now()
+	dialer, ok := d.dialer.(proxy.ContextDialer)
+	if !ok {
+		return nil, errors.New("unable to cast socks proxy dialer to context proxy dialer")
+	}
+	c, err := dialer.DialContext(ctx, n, addr)
+
+	var code string
+	if err == nil {
+		code = "0"
+	} else if socksErr, ok := err.(*net.OpError); ok && strings.Contains(socksErr.Op, "socks") {
+		// Socks errors defined here: https://cs.opensource.google/go/x/net/+/refs/tags/v0.15.0:internal/socks/socks.go;l=40-63
+		if strings.Contains(socksErr.Error(), "general SOCKS server failure") {
```

Comment:
A shame the `code` isn't available as part of `socksErr` here to avoid the string comparison 😞

Reply:
💯 This would have made everything easier.